tree/ntuple/v7/src/RFieldMeta.cxx

```diff
@@ -197,6 +198,37 @@ void ROOT::Experimental::RClassField::ReadInClusterImpl(RClusterIndex clusterInd
    }
 }
 
+void ROOT::Experimental::RClassField::BeforeConnectPageSource(Internal::RPageSource &pageSource)
+{
+   if (GetOnDiskId() == kInvalidDescriptorId) {
+      return;
+   }
+   // Gather all known sub fields in the descriptor.
+   std::unordered_set<std::string> knownSubFields;
+   {
+      const auto descriptorGuard = pageSource.GetSharedDescriptorGuard();
+      const RNTupleDescriptor &desc = descriptorGuard.GetRef();
+      const auto &fieldDesc = desc.GetFieldDescriptor(GetOnDiskId());
+      // Check that we have the same type; only then we perform automatic schema evolution.
+      // TODO: should this throw? Right now, we allow reading back a field hierarchy with a different type as long as
+      // all columns match...
+      if (GetTypeName() != fieldDesc.GetTypeName())
+         return;
```

Comment:
That's a question to discuss...

Reply:
That's a good question! Perhaps leave it in here for now to avoid really confusing conditions for the time being. But it remains a TODO. I think we may generally need to think about model compatibility (structural compatibility). But that would then be checked at a different place.

Reply:
As discussed offline, we generally forbid the case of mismatching type names of classes for the moment.